flintrock/core.py

```diff
@@ -506,42 +507,73 @@ def get_java_major_version(client: paramiko.client.SSHClient):
                 client=client,
                 command=command)
             tokens = output.split()
-            # First line of the output is like: 'java version "1.8.0_20"'
+            # First line of the output is like: 'openjdk version "1.8.0_252"' or 'openjdk version "11.0.7" 2020-04-14'
             # Get the version string and strip out the first two parts of the
-            # version as a tuple: (1, 8)
+            # version as an int: 7, 8, 9, 10...
             if len(tokens) >= 3:
                 version_parts = tokens[2].strip('"').split(".")
                 if len(version_parts) >= 2:
-                    return tuple(int(part) for part in version_parts[:2])
+                    if version_parts[0] == "1":
+                        # Java 6, 7 or 8
+                        return int(version_parts[1])
+                    else:
+                        # Java 9+
+                        return int(version_parts[0])
         except SSHError:
             pass
 
     return None
 
 
-def ensure_java8(client: paramiko.client.SSHClient):
+def ensure_java(client: paramiko.client.SSHClient, java_version: int):
     host = client.get_transport().getpeername()[0]
     java_major_version = get_java_major_version(client)
 
-    if not java_major_version or java_major_version < (1, 8):
-        logger.info("[{h}] Installing Java 1.8...".format(h=host))
+    if java_major_version and java_major_version == java_version:
+        logger.info("Java {j} is already installed, skipping Java install".format(j=java_major_version))
+    elif java_major_version and java_major_version > java_version:
+        logger.warning("Downgrading existing Java {j} installation".format(j=java_major_version))
 
+    # sudo yum install -y java-1.8.0-openjdk
+    # sudo amazon-linux-extras install -y java-openjdk11
+    elif not java_major_version or java_major_version < java_version:
+        logger.info("[{h}] Installing Java {j}...".format(h=host, j=java_version))
```

Comment:
I think here you probably just want to have something like:

```python
else:
    if java_major_version > java_version:
        logger.warning("Downgrading existing Java {j} installation".format(j=java_major_version))
    logger.info("[{h}] Installing Java {j}...".format(h=host, j=java_version))
    ...
```

What you have now will show a warning if the existing Java is newer than what's requested, but it won't actually downgrade the version.